ADD RSpec/SharedExamples cop

anthony-robin opened the pull request:
Enforces use of string to titleize shared examples.

@example
```ruby
  # bad
  it_behaves_like :foo_bar_baz
  it_should_behave_like :foo_bar_baz
  shared_examples :foo_bar_baz
  shared_examples_for :foo_bar_baz
  include_examples :foo_bar_baz

  # good
  it_behaves_like 'foo bar baz'
  it_should_behave_like 'foo bar baz'
  shared_examples 'foo bar baz'
  shared_examples_for 'foo bar baz'
  include_examples 'foo bar baz'
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) is passing.

If you have created a new cop:
* [x] Added the new cop to `config/default.yml`.
* [x] The cop includes examples of good and bad code.
* [x] You have tests for both code that should be reported and code that is good.

bquorning commented:
Why do you prefer `string` over `symbolic` as the default config?

bquorning commented on lib/rubocop/cop/rspec/shared_examples_style.rb:
```diff
@@ -0,0 +1,136 @@
+          end
+
+          def preferred_style
+            ":#{string.tr(' ', '_').tr('"', '')}"
```
The substitution of `”` could be moved into `#string`?

bquorning commented on lib/rubocop/cop/rspec/shared_examples_style.rb:
```diff
@@ -0,0 +1,136 @@
+          end
+
+          def preferred_style
+            "'#{symbol.to_s.tr('_', ' ')}'"
```
Perhaps extract a method that wraps a given string with single quotes?

bquorning commented:
If you’re wondering about the red builds… #567 should fix them.

bquorning commented:
Please rebase on master.

anthony-robin commented:
> Why do you prefer string over symbolic as the default config?

I chose string as I think it makes shared examples more readable than symbols but I can change it. 
Anyway, both configuration add offenses for already existing shared examples. Should I fix them in this PR or in a separate one according to the default style ?

bquorning commented:
I’m looking through [the documentation](https://relishapp.com/rspec/rspec-core/v/3-6/docs/example-groups/shared-examples), and [the source code](https://github.com/rspec/rspec-core), and in both places, only string arguments are used.

We should discuss the option of only allowing string arguments to these methods.

bquorning commented on lib/rubocop/cop/rspec/shared_examples_style.rb:
```diff
@@ -0,0 +1,136 @@
+          (send nil? {
+              :it_behaves_like :it_should_behave_like
+              :shared_examples :shared_examples_for
+              :include_examples
```
I think we may have some useful constants available in the `RuboCop::RSpec::Language` namespace.

bquorning commented:
> […] offenses for already existing shared examples. Should I fix them in this PR or in a separate one according to the default style ?

Yes please, fix them (in a separate commit) in this PR.

anthony-robin commented:
> We should discuss the option of only allowing string arguments to these methods.

I agree, only allowing strings would be better.

anthony-robin commented:
I just discovered that the pattern does not detect examples with several arguments such as:

```
include_examples :explicit_autocorrect, 'be_truthy', 'be_falsey'
```

What did I miss ?

anthony-robin commented on lib/rubocop/cop/rspec/shared_examples_style.rb:
```diff
@@ -0,0 +1,80 @@
+        end
+
+        # :nodoc:
+        class StringStyleChecker
```
I kept the class as I think it makes things easier to understand but if you think it is not necessary anymore since there is only one style allowed I can change it.

bquorning replied:
Perhaps the class name is too verbose now that there is just 1 style. How do you like `StyleChecker` or just `Checker`?

bquorning replied:
I kind of like this logic encapsulated in a class, even if there is only one style.

bquorning commented on lib/rubocop/cop/rspec/shared_examples_style.rb:
```diff
@@ -0,0 +1,80 @@
+      #   shared_examples_for 'foo bar baz'
+      #   include_examples 'foo bar baz'
+      #
+      class SharedExamplesStyle < Cop
```
Perhaps the “Style” suffix is unnecessary?

bquorning commented on lib/rubocop/cop/rspec/shared_examples_style.rb:
```diff
@@ -0,0 +1,80 @@
+      #
+      class SharedExamplesStyle < Cop
+        def_node_matcher :shared_examples_style, <<-PATTERN
+          (send nil? {#{(SharedGroups::ALL + Includes::ALL).node_pattern}} $sym)
```
We want to capture everything that isn’t a string, right? So I think it should be `$!str` instead of `$sym`.

bquorning commented on lib/rubocop/cop/rspec/shared_examples_style.rb:
```diff
@@ -0,0 +1,80 @@
+          end
+
+          def offensive?
+            !node.str_type?
```
This method will always return true, since `node` is (currently) always a string node. You can remove the `#offensive?` method altogether.

bquorning commented:
> What did I miss ?

You missed a failing test case 😉

You can add `...` to the end of the node pattern to allow the extra arguments.

anthony-robin commented:
I refactored code according to your review. I let the `offensive?` method as I think it is necessary now.

Thanks !

bquorning commented on lib/rubocop/cop/rspec/shared_examples.rb:
```diff
@@ -0,0 +1,80 @@
+        def on_send(node)
+          shared_examples(node) do |ast_node|
+            checker = Checker.new(ast_node)
+            return unless checker.offensive?
```
This block is only executed when an `ast_node` is found that matches the criteria `!str`. This means `ast_node.str_type?` will always be false, and `checker.offensive?` will always be true.

anthony-robin replied:
Right, I got it now !

bquorning commented:
So, I just checked this, and… the valid argument types to `shared_examples` etc. are `String`, `Symbol` and `Module`. 

https://github.com/rspec/rspec-core/blob/6fdcbfec431a918afedeabcae17532fb0b1b2c2d/lib/rspec/core/shared_example_group.rb#L206-L211

I managed to dig 7 years back to this pr: https://github.com/rspec/rspec-core/pull/408

anthony-robin commented:
> the valid argument types to `shared_examples` etc. are `String`, `Symbol` and `Module`.

Do you want me to handle all of this types as before my refactoring and allow users to configure the style they want or we still enforce usage of string ?

bquorning commented:
I believe `Module`/`Class` arguments cannot be easily substituted for a string/symbol equivalent. And vice versa. The best we can do it probably to *discourage* using symbols instead of insisting on *always* using strings.

So you were right earlier, to have `$sym` in your node matcher instead of the `$!str` that I suggested.

Darhazer commented on spec/rubocop/cop/rspec/shared_examples_spec.rb:
```diff
@@ -0,0 +1,52 @@
+                      ^^^^^^^^^^^^ Prefer 'foo bar baz' over `:foo_bar_baz` to titleize shared examples.
+      it_should_behave_like :foo_bar_baz
+                            ^^^^^^^^^^^^ Prefer 'foo bar baz' over `:foo_bar_baz` to titleize shared examples.
+      shared_examples :foo_bar_baz
```
can you test those with blocks (shared_examples :foo_bar_baz do |param|)?

anthony-robin commented:
I refactored my code to only detect offenses for symbols titles.
I also added tests for shared examples with blocks and shared examples with modules/class name as title to ensure they don't throw offenses.

Thanks !

bquorning commented:
@Darhazer I would to head your opinion on this cop, too.

Darhazer commented:
I've never seen usage of symbol for shared_example title, but the fact that our own code base had such examples shows that it might be useful cop
